storage/storage_dest.go

```diff
@@ -317,122 +325,141 @@ func (s *storageImageDestination) TryReusingBlobWithOptions(ctx context.Context,
+				layers, err := s.imageRef.transport.store.LayersByUncompressedDigest(uncompressedDigest)
+				if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
+					return false, private.ReusedBlob{}, fmt.Errorf(`looking for layers with digest %q: %w`, uncompressedDigest, err)
+				}
+				if len(layers) > 0 {
+					if size != -1 {
+						s.blobDiffIDs[digestBlob] = layers[0].UncompressedDigest
+						return true, private.ReusedBlob{
+							Digest: digestBlob,
+							Size:   size,
+						}, nil
+					}
+					if !options.CanSubstitute {
+						return false, private.ReusedBlob{}, fmt.Errorf("Internal error: options.CanSubstitute was expected to be true for blob with digest %s", digestBlob)
+					}
+					s.blobDiffIDs[uncompressedDigest] = layers[0].UncompressedDigest
 					return true, private.ReusedBlob{
-						Digest: digest,
-						Size:   size,
+						Digest: uncompressedDigest,
+						Size:   layers[0].UncompressedSize,
 					}, nil
 				}
-				if !options.CanSubstitute {
-					return false, private.ReusedBlob{}, fmt.Errorf("Internal error: options.CanSubstitute was expected to be true for blob with digest %s", digest)
-				}
-				s.uncompressedOrTocDigest[uncompressedDigest] = layers[0].UncompressedDigest
-				return true, private.ReusedBlob{
-					Digest: uncompressedDigest,
-					Size:   layers[0].UncompressedSize,
-				}, nil
 			}
 		}
 	}
 
-	tocDigest := digest
-	if options.TOCDigest != nil {
-		tocDigest = *options.TOCDigest
-	}
+	if options.TOCDigest != "" {
+		// we are already holding s.lock
+		diffOutput, ok := s.diffOutputs[digestBlob]
+		if ok {
+			// Save this for completeness.
+			s.tocDigests[options.TOCDigest] = options.TOCDigest
+			return true, private.ReusedBlob{
+				TOCDigest: diffOutput.TOCDigest,
+				Size:      diffOutput.Size,
+			}, nil
+		}
 
-	// Check if we have a chunked layer in storage with the same TOC digest.
-	layers, err = s.imageRef.transport.store.LayersByTOCDigest(tocDigest)
-	if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
-		return false, private.ReusedBlob{}, fmt.Errorf(`looking for layers with TOC digest %q: %w`, tocDigest, err)
-	}
-	if len(layers) > 0 {
-		// Save this for completeness.
-		s.uncompressedOrTocDigest[digest] = layers[0].TOCDigest
-		return true, private.ReusedBlob{
-			Digest: layers[0].TOCDigest,
-			Size:   layers[0].UncompressedSize,
-		}, nil
+		// Check if we have a chunked layer in storage with the same TOC digest.
+		layers, err := s.imageRef.transport.store.LayersByTOCDigest(options.TOCDigest)
+
+		if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
+			return false, private.ReusedBlob{}, fmt.Errorf(`looking for layers with TOC digest %q: %w`, options.TOCDigest, err)
+		}
+		if len(layers) > 0 {
+			// Save this for completeness.
+			s.tocDigests[options.TOCDigest] = layers[0].TOCDigest
+			return true, private.ReusedBlob{
+				TOCDigest: layers[0].TOCDigest,
```

Comment:
What digest will be written in the manifest if `Digest` is unknown?

Yes, c/image/storage currently lies here because it wants to preserve the original manifest (and sometimes fails in that)

Reply:
in `LayerInfosForCopy()` we override the digest with the expected-digest, that was stored here: https://github.com/containers/image/blob/main/storage/storage_dest.go#L672-L675

the expected-digest is not validated when we pull the image but it is validated when we copy from the storage.

The TOC instead is validated by `ApplyDiffWithDiffer`.  So if the TOC is not valid, `PutBlobPartial` fails.

So we lookup a layer by its TOC only if the TOC was specified in an annotation.  An image can use a layer by its TOC only if the image manifest contains the same TOC digest.  We do not lookup by TOC if the layer has no TOC annotation.

When we copy from the storage, we override the Digest with the expected-digest flag, but this is going to be validated later.

A layer that doesn't have a TOC annotation must not use a layer from the storage if it is stored by its TOC.

Reply:
I agree that `LayerInfosForCopy` handles the “push” use case fine.

But, still, a “pull” does record a manifest in c/storage.

In general the destination transport is supposed to collaborate with the generic copy code by returning the right partial edits in `ReusedBlob`/`UploadedBlob`, so that the destination can receive an “appropriate” manifest that works correctly with the blobs.

What we _really_ want to achieve with that in c/storage is to record the original, unmodified, version of the manifest, along with signatures, so that when the image is found to contain a backdoor, we can ~prove it was originally correctly signed. (That doesn’t always work when pulling encrypted images, but that’s the goal.)

Given that context, I think the `ReusedBlob` data returned here should be designed so that it doesn’t trigger any manifest update.

Also, the compressed-blob digest is _mandatory_ in manifests, so we can’t very well store a manifest in c/storage which just doesn’t _have_ that field. We could, sure, decide that `ReusedBlob` can return no `Digest` field, and in that case the generic copy pipeline will not modify the value. Given the special treatment of manifest/blob updates in c/storage, I think keeping that special case inside c/storage is a bit more appropriate.

(I could also well accept an argument that doing all this to store the original manifest in c/storage is just wrong, and that we should record an “updated” manifest in c/storage, and then we can get ~rid of `LayerInfosForCopy`; and we should have a _separate“ `PutOriginalManifest` method to deal with the desired feature that way. But that’s not what we have, and `LayerInfosForCopy` needs to continue to be supported as part of the stable API.)

Reply:
This was already updated to preserve `Digest: blobDigest`.